core/pow/engine.go

```diff
@@ -53,18 +55,29 @@ func (p *params) isActive(blockHeight uint64) bool {
 	return p.untilBlock == nil || *p.untilBlock+p.spamPoWNumberOfPastBlocks > blockHeight
 }
 
+// represents the number of transactions seen from a party and the total observed difficulty
+// of transactions generated with a given block height.
+type partyStateForBlock struct {
+	observedDifficulty uint
+	seenCount          uint
+}
+
 type state struct {
-	blockPartyToObservedDifficulty map[string]uint // party observed total difficulty in block
-	blockPartyToSeenCount          map[string]uint // party observed transactions in block
+	blockToPartyState map[uint64]map[string]*partyStateForBlock
```

Comment:
Since maps in golang don't shrink automatically in memory, I think we should find a way to clean up the maps. Like recreating a new one when appropriate, for example ?

I don't know what would be best in terms of technic.

Reply:
we clean up the state when possible, i.e. when a block becomes out of scope we remove it from blockToPartyState. Otherwise it's add only because all information is needed while the block is still in scope.